### ISSUE TITLE
LCD Z-fighting

### ISSUE THREAD

**Issue by VanessaE** [minetest-mods/digilines#56]
When you place and use an LCD, it works, but if you go away for a bit, then return to the area, forcing the text entity to re-load, its position is now too close to (or flush with) the LCD surface, causing Z-fighting or making the text entirely invisible:

![Screenshot_2019-09-09_16-22-25](https://user-images.githubusercontent.com/1733537/64563719-081b5780-d31e-11e9-9a01-14fa32dc9353.png)

I had similar problems with signs_lib at one time.  The solution was to simply move the entity out a bit further from the sign.  Far enough away to avoid Z-fighting, but still close enough that you have to get your nose right up on it to notice it isn't flush.

**Comment by DS-Minetest** [comment]
Related: #40

I can reproduce z-fighting (without reloading the object) very far away (at 30000).

**Comment by numberZero** [comment]
From #40:

> this should ideally be tested on various gfx cards.

So 0.5mm appears to be *too* near.

https://github.com/minetest-mods/digilines/blob/2d2e77899409ba36aed834f0ce9c2c534f17b3d2/lcd.lua#L23-L33

Change 0.437 to something reasonable, like 0.435 or 7/16−1/256 (current value is 7/16−1/2000, previous value was 7/16−3/80), and tell us what values look good. Z-fighting is unacceptable ofc, but large gap is unpleasant too.

**Comment by VanessaE** [comment]
Lines 111-114 in current code;  https://github.com/minetest-mods/digilines/blob/master/lcd.lua#L111-L114

They do not have to be an even multiple of anything, the positions are entirely free-form.

I changed them to **±0.42**.  That seems to be enough to prevent Z-fighting at normal viewing distance after re-load, and it still fades away as you walk away.

I don't know if the position being wrong is the engine's fault, the mod's fault, or irrlicht's, but I'm pretty sure it'll be the same from one GPU to another.

**Comment by DS-Minetest** [comment]
#57

**Comment by DS-Minetest** [comment]
I've tested the fading away. For me (and with the PR) it happens at a distance of ca. 20 nodes from the lcd, regardless of the position in the world. I guess the reason for this is the bit-size of the depth buffer.

**Comment by DS-Minetest** [comment]
I've tested the fading away again without mipmapping and it was barely noticable. I have to admit that I'm stupid.

**Comment by numberZero** [comment]
Works for me, Minetest 5.1.0-dev-429a98964 (Linux). What are your settings? (and GPU vendor)

**Comment by VanessaE** [comment]
In the case of signs_lib, I fixed this issue once and for all today by not storing the entity statically, regenerating it with with an LBM, deleting any entities that may have been there at load time, before spawning a new one (there already was an LBM in place for replacing entities lost due to `/clearobjects`, so I just expanded it's purpose slightly).  It's fast and less error-prone.

https://gitlab.com/VanessaE/signs_lib/commit/4b2abfadce911c26ba14e318ee4f0a6a3b687430

This, plus a few subsequent commits to clean up some leftovers and fix a couple tiny bugs that it caused.

The signs' entities all use a mesh model consisting of a single rectangle (so two triangles), with the entity's origin at the node center.

**Comment by VanessaE** [comment]
(I realize I dropped support for the old 7px font during the recent rewrite, and need to re-add it should the LCD be altered to use signs_lib, but I may just add it anyway)

**Comment by VanessaE** [comment]
In the case of signs_lib, I hopefully fixed this issue once and for all today by not storing the entity statically, instead regenerating it with with an LBM, deleting any entities that may have been there at load time, before spawning a new one (there already was an LBM in place for replacing entities lost due to `/clearobjects`, and all entities' textures have to be fully re-rendered on-load anyways regardless of how the entity got there, so I just expanded the LBM's purpose slightly).  It's fast and less error-prone than Minetest's own on-load entity handling.

https://gitlab.com/VanessaE/signs_lib/commit/4b2abfadce911c26ba14e318ee4f0a6a3b687430

This commit did the bulk of it, plus a few subsequent commits just to clean up some leftovers and fix a couple tiny bugs that it caused.

The signs' entities all use a mesh model consisting of a single rectangle (so two triangles), with the model's origin at the upper left corner of the rectangle, and the entity's origin at the node center.  The rectangle is UV-mapped to fit the image, so no special offsetting or clipping of the texture is needed.

I suggest doing the same with the LCD, or perhaps it's possible to use signs_lib to generate the display?  I realize I dropped support for the old 7px font during the recent rewrite, and would need to re-add it should the LCD be altered to use signs_lib, but I may just add it anyway, if only for future use somewhere...
